src/main/java/sparksoniq/jsoniq/runtime/iterator/control/SwitchRuntimeIterator.java

```diff
@@ -44,37 +48,28 @@ public boolean hasNext() {
     }
 
     @Override
-    public void reset(DynamicContext context){
+    public void reset(DynamicContext context) {
         super.reset(context);
         this.matchingIterator = null;
         initializeIterator(testField, cases, defaultReturn);
     }
 
     private void initializeIterator(RuntimeIterator test, Map<RuntimeIterator, RuntimeIterator> cases,
                                     RuntimeIterator defaultReturn) {
-        test.open(_currentDynamicContext);
-        Item testValue = test.next();
-        if(test.hasNext())
-            throw new NonAtomicKeyException("Switch test must be atomic", getMetadata().getExpressionMetadata());
-        test.close();
-        for(RuntimeIterator caseKey : cases.keySet()){
-            caseKey.open(_currentDynamicContext);
-            Item caseValue = caseKey.next();
-            if(caseKey.hasNext())
-                throw new NonAtomicKeyException("Switch case test must be atomic", getMetadata().getExpressionMetadata());
+        Item testValue = getSingleItemOfTypeFromIterator(test, Item.class,
+                new NonAtomicKeyException("Switch test must be atomic", getMetadata().getExpressionMetadata()));
```

Comment:
Passing only the string rather than an exception object would probably make the code faster, because other the exception object is created every time, even when there is no exception to throw.